### PR TITLE
[ITensors] [ENHANCEMENT] Allow `Matrix` representations for operators in `expect` and `correlation_matrix`

### PR DESCRIPTION
**Pull request by mtfishman**
# Description

Allow `Matrix` representations for operators in `expect` and `correlation_matrix`. This can make it a bit easier to measure an operator that is not defined already without having to define the operator by overloading `op`.

For inspiration see the discussion here: https://itensor.discourse.group/t/measuring-correlations/51/6

<details><summary>Minimal demonstration of new behavior</summary><p>

```julia
julia> using ITensors

julia> s = siteinds("S=1/2", 4);

julia> ψ = randomMPS(s);

julia> expect(ψ, [1/2 0; 0 -1/2])
4-element Vector{Float64}:
  0.1968928836524881
 -0.4253469177197064
  0.4028392528567332
 -0.4026126134149185

julia> expect(ψ, "Sz")
4-element Vector{Float64}:
  0.1968928836524881
 -0.4253469177197064
  0.4028392528567332
 -0.4026126134149185

julia> correlation_matrix(ψ, [1/2 0; 0 -1/2], [1/2 0; 0 -1/2])
4×4 Matrix{Float64}:
  0.25       -0.0837478   0.0793162  -0.0792716
 -0.0837478   0.25       -0.171346    0.17125
  0.0793162  -0.171346    0.25       -0.162188
 -0.0792716   0.17125    -0.162188    0.25

julia> correlation_matrix(ψ, "Sz", "Sz")
4×4 Matrix{Float64}:
  0.25       -0.0837478   0.0793162  -0.0792716
 -0.0837478   0.25       -0.171346    0.17125
  0.0793162  -0.171346    0.25       -0.162188
 -0.0792716   0.17125    -0.162188    0.25
```
</p></details>

# How Has This Been Tested?

Added tests of the behavior shown above.

# Checklist:

- [x] My code follows the style guidelines of this project. Please run `using JuliaFormatter; format(".")` in the base directory of the repository (`~/.julia/dev/ITensors`) to format your code according to our style guidelines.
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have added tests that verify the behavior of the changes I made.
- [x] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [x] Any dependent changes have been merged and published in downstream modules.

### CODE REVIEW

**Review by emstoudenmire — APPROVED** [review]
Looks good. Not a big change to the code required.

**Comment by emstoudenmire** [comment]
I like the idea so let's merge this. The only possible failure modes are these:

1. for fermions, making the operator from a matrix will lose the information about whether it has J-W string in the current system, so would lead to wrong results if someone tried to make "C" or "Cdag" from a matrix
2. if the sites are of mixed type or dimension, then the operator from a matrix may only be meaningful for some of the sites

I'm ok with #1 because it's just another one of the pitfalls of the current fermion system, and would be fixed by the new one. For #2 it's a little more of a possible silent failure mode if all the sites have the same dimension but a different physical meaning. But I think we're ok there because the user is still being explicit about what they are doing i.e. "sandwich this matrix between every site" so they have to know what to expect.

In summary again, I think this is good to merge.

**Comment by mtfishman** [comment]
Agreed that those are the main downsides, but I also agree with your assessment that 1. should be fixed by the new Fermion system and 2. is acceptable for the reasons you give.
